zerver/views/streams.py

```diff
@@ -664,22 +664,22 @@ def add_subscriptions_backend(
         realm, streams, subscribers, acting_user=user_profile, color_map=color_map
     )
 
-    # We can assume unique emails here for now, but we should eventually
-    # convert this function to be more id-centric.
-    email_to_user_profile: dict[str, UserProfile] = {}
+    id_to_user_profile: dict[str, UserProfile] = {}
 
     result: dict[str, Any] = dict(
         subscribed=defaultdict(list), already_subscribed=defaultdict(list)
     )
     for sub_info in subscribed:
         subscriber = sub_info.user
         stream = sub_info.stream
-        result["subscribed"][subscriber.email].append(stream.name)
-        email_to_user_profile[subscriber.email] = subscriber
+        user_id = str(subscriber.id)
+        result["subscribed"][user_id].append(stream.name)
+        id_to_user_profile[user_id] = subscriber
     for sub_info in already_subscribed:
         subscriber = sub_info.user
         stream = sub_info.stream
-        result["already_subscribed"][subscriber.email].append(stream.name)
+        user_id = str(subscriber.id)
```

Comment:
Why do these need to be stored as strings rather than integers?

Reply:
The results are to be returned in JSON format, which is why these are stored as strings rather than integers.